### ISSUE TITLE
[Discussion] Documentation improvements for MTIAsyncVideoCompositionRequestHandler

### ISSUE THREAD

**Issue by samuelhorwitz** [MetalPetal/MetalPetal#236]
### Checklist
- [x] I've read the [README](https://github.com/MetalPetal/MetalPetal/blob/master/README.md)
- [x] I've searched for [existing GitHub issues](https://github.com/MetalPetal/MetalPetal/issues)

First of all, this is a great project, it saved me from shutting down an idea I had while still in the exploration phase because without this project, there doesn't seem to be a great way to do on-line video compositing in the AV framework. This was exactly what I needed.

# Discussion

Long story short, I have been dealing with frequent deadlocking that was tied to `videoComposition` on `AVPlayerItem`. It seemed to occur when player items were being discarded (`AVQueuePlayer` queue replacements) and when `videoComposition` was getting a new composition plugged into it. Tactics that somewhat alleviated things (made the deadlocking slightly less common) was assigning `videoComposition` on a global dispatch queue as well as pausing playback before making `videoComposition` changes as well as debouncing user input that led to these changes and debouncing playback, itself.

Everything pointed to some kind of rendering deadlocking when an existing `videoComposition` was about to be torn down whether directly through that parameter being set to a new value, or through the deallocation of an `AVPlayerItem`.

I was going to again dig into it some more today when I finally noticed `MTIAsyncVideoCompositionRequestHandler` takes an optional `queue` parameter that defaults to `.main`. I had completely missed this as that line of code in my project had not really changed much since I had begun to prototype things. I changed it to a global queue (specifically `.global(qos: .userInteractive)` but it probably does not matter which one) and just like that, all the deadlocking I had been encountering is seemingly fixed. The deadlocking was extremely easy for me to reproduce; all I had to do was make some of the user input buttons completely un-debounced and rapidly click them and within seconds, the app would hang. So I am confident in saying the deadlocking issue was solved by this. It is not a rare one-off issue, it was something plaguing my app on almost every run.

Now that may be obvious, of course why wouldn't I look at something very close to the source of the deadlock that was explicitly executing on the main thread? Well I would have if I had known about it earlier :D. That's what this issue is really about.

## tl;dr

I think at the very least the README should perhaps highlight that `MTIAsyncVideoCompositionRequestHandler` takes a queue parameter. I'd also question why it has a default value and why that default value is `.main`. The deprecation issues might make it infeasible to change the API but I think it may make sense if queue has to be specified by the user to make intent very explicit. But even if no API changes are made, I really think a highlighted section in the README could help. Also, I hope this issue being here helps anyone searching "deadlock" at some future point.

I am attaching two videos to demonstrate this as I know that saying something is caused by deadlocking or race conditions can be sometimes a bit hand-wavy and suspicious. In the two videos below I have completely removed all user input debouncing so that each button click triggers updates (specifically, it assigns a new `videoComposition` to an `AVPlayerItem` each click). In case it's not clear, I'm rapid fire clicking the button in both videos and in the first one it almost immediately locks up.

## Deadlocking (`MTIAsyncVideoCompositionRequestHandler` using `.main`)

https://user-images.githubusercontent.com/1251501/112840067-e6eb9a80-906c-11eb-836d-10948f210ac4.mp4

## No deadlocking (`MTIAsyncVideoCompositionRequestHandler` using `.global(qos: .userInteractive)`)

https://user-images.githubusercontent.com/1251501/112840723-ad675f00-906d-11eb-857b-276e32b3c83b.mp4

**Comment by YuAo** [comment]
Thanks for the report!

I tried to reproduce the issue, and yes, callback on the main queue may cause a deadlock on macOS (condition: change the playerItem's videoComposition duration playback and the videoComposition's renderSize must be a different value), however deadlock is not observed with the same configuration on iOS.

We use main queue as the default queue because framework users are likely to use filter related things in the `filter` block, and filters are not thread-safe. It is common to change a filter's parameter on main thread (e.g. when a user sliding a slider).

I think I need to take a closer look at this before coming up with a proper solution. Thanks again for the detailed report!

**Comment by YuAo** [comment]
The underlying problem is related to the cancelling mechanism of the video compositor. It seems that the video compositor on macOS still waits for the `asyncVideoCompositionRequest` to finish when the request is cancelled.

- `AVPlayerItem.currentTime` blocks the main queue, waiting for a lock that the video compositing system is holding.
- The video compositing system cannot release the lock because it is waiting for cancelled `asyncVideoCompositionRequest`s to finish.
- Cancelled `asyncVideoCompositionRequest`s cannot finish because the request handler needs the main queue to run and unfortunately, the main queue is blocked by  `AVPlayerItem.currentTime`. 

So a better solution here is to implement the `cancelAllPendingVideoCompositionRequests` method in our `BlockBasedVideoCompositor` and actually cancel all pending requests.

**Comment by samuelhorwitz** [comment]
Thanks for the follow up, that sounds about right! I wasn’t sure if I was just approaching things incorrectly so I flagged as documentation but I guess it’s a deeper issue. Either way this solution unblocks me for the time being as my approach with filter changes is to rebuild the whole `AVVideoComposition` instance versus mutating filters. Perhaps that’s a thing for me to look into optimizing on my end eventually now that you point out that flow, but either way thanks for digging into this also!
